sensor/common/deduper/deduper.go

```diff
@@ -58,24 +61,35 @@ var (
 
 // deduper takes care of deduping sensor events.
 type deduper struct {
-	stream   messagestream.SensorMessageStream
-	lastSent map[Key]uint64
+	stream         messagestream.SensorMessageStream
+	lastSent       map[deduperkey.Key]uint64
+	observationSet *ClosableSet
 
 	hasher *hash.Hasher
 }
 
 // NewDedupingMessageStream wraps a SensorMessageStream and dedupes events. Other message types are forwarded as-is.
-func NewDedupingMessageStream(stream messagestream.SensorMessageStream, deduperState map[Key]uint64) messagestream.SensorMessageStream {
+func NewDedupingMessageStream(stream messagestream.SensorMessageStream, deduperState map[deduperkey.Key]uint64, observationSet *ClosableSet) messagestream.SensorMessageStream {
 	if deduperState == nil {
-		deduperState = make(map[Key]uint64)
+		deduperState = make(map[deduperkey.Key]uint64)
 	}
 	return &deduper{
-		stream:   stream,
-		lastSent: deduperState,
-		hasher:   hash.NewHasher(),
+		stream:         stream,
+		lastSent:       deduperState,
+		hasher:         hash.NewHasher(),
+		observationSet: observationSet,
 	}
 }
 
+func buildKey(typ, id string) string {
+	return fmt.Sprintf("%s:%s", typ, id)
+}
+
+func getKey(msg *central.MsgFromSensor) string {
+	event := msg.GetEvent()
+	return buildKey(eventPkg.GetEventTypeWithoutPrefix(event.GetResource()), event.GetId())
+}
```

Comment:
[nit] This functions are also defined [here](https://github.com/stackrox/stackrox/blob/4ba873b685b734595366b78880c6c2af8ada9e9d/central/hash/manager/deduper.go#L124-L131). Would it make sense to maybe extract them in a common package under `pkg`?

Reply:
Created a follow-up for this: [ROX-20693](https://issues.redhat.com/browse/ROX-20693)
I think there might be other things we can pull up to a shared package.